consensus/hotstuff/signature/packager.go

```diff
@@ -1,26 +1,238 @@
 func (p *ConsensusSigPackerImpl) Unpack(blockID flow.Identifier, signerIDs []flow.Identifier, sigData []byte) (*hotstuff.BlockSignatureData, error) {
-	panic("to be implemented")
+	// decode into typed data
+	var data signatureData
+	err := p.encoder.Decode(sigData, &data)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode sig data %s: %w", err, signature.ErrInvalidFormat)
+	}
+
+	// deserialize the compact sig types
+	sigTypes, err := deserializeFromBytes(data.SigType, len(signerIDs))
+	if err != nil {
+		return nil, fmt.Errorf("failed to deserialize sig types from bytes: %w", err)
+	}
+
+	// read all the possible signer IDs at the given block
+	consensus, err := p.committees.Identities(blockID, filter.HasRole(flow.RoleConsensus))
+	if err != nil {
+		return nil, fmt.Errorf("could not find consensus committees by block id(%v): %w", blockID, err)
+	}
+
+	// lookup is a map from node identifier to node identity
+	// it is used to check the given signerIDs are all valid signers at the given block
+	lookup := consensus.Lookup()
+
+	// read each signer's signerID and sig type from two different slices
+	// group signers by its sig type
+	stakingSigners := make([]flow.Identifier, 0)
+	randomBeaconSigners := make([]flow.Identifier, 0)
+
+	for i, sigType := range sigTypes {
+		signerID := signerIDs[i]
+		_, ok := lookup[signerID]
+		if !ok {
+			return nil, fmt.Errorf("unknown signer ID (%v) at the given block (%v): %w",
+				signerID, blockID, signature.ErrInvalidFormat)
+		}
+
+		if sigType == hotstuff.SigTypeStaking {
+			stakingSigners = append(stakingSigners, signerID)
+		} else if sigType == hotstuff.SigTypeRandomBeacon {
+			randomBeaconSigners = append(randomBeaconSigners, signerID)
+		} else {
+			return nil, fmt.Errorf("unknown sigType %v, %w", sigType, signature.ErrInvalidFormat)
+		}
+	}
+
+	return &hotstuff.BlockSignatureData{
+		StakingSigners:               stakingSigners,
+		RandomBeaconSigners:          randomBeaconSigners,
+		AggregatedStakingSig:         data.AggregatedStakingSig,
+		AggregatedRandomBeaconSig:    data.AggregatedRandomBeaconSig,
+		ReconstructedRandomBeaconSig: data.RandomBeacon,
+	}, nil
+}
+
+// serializeToBytes encodes the given sigTypes into a bit vector.
+// We append tailing `0`s to the vector to represent it as bytes.
+func serializeToBytes(sigTypes []hotstuff.SigType) ([]byte, error) {
+	bytes := make([]byte, 0)
+	// a sig type can be converted into one bit.
+	// so every 8 sig types will fit into one byte.
+	// iterate through every 8 sig types, for each sig type in each
+	// group, convert into each bit, and fill the bit into the byte.
+	// the remaining unfilled bits in the last byte will be 0
+	for byt := 0; byt < len(sigTypes); byt += 8 {
+		b := byte(0)
+		offset := 7
+		for pos := byt; pos < byt+8 && pos < len(sigTypes); pos++ {
+			sigType := sigTypes[pos]
+
+			if !sigType.Valid() {
+				return nil, fmt.Errorf("invalid sig type: %v at pos %v", sigType, pos)
+			}
+
```

Comment:
I agree the code would be future resilient, but I don't like that the logic here knows that `SigTypeStaking` is for `0`, and `SigTypeRandomBeacon` is for `1`. 

The reason I used:
```
sigType := sigTypes[pos]
b ^= (byte(sigType) << offset)
``` 
is so that the value is determined by the iota value defined in the type. So if we change the order of types in `SigType` definition, this code doesn't need to change.

If we need to introduce more type, we anyway will need to update the code here. And if we don't, both your approach and mine will throw the same exception.